src/service/short_url.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 Zinc Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use chrono::Utc;
+use config::{get_config, utils::md5};
+use infra::{
+    errors::{DbError, Error},
+    short_url::ShortUrlRecord,
+};
+
+use crate::service::db;
+
+pub fn get_base_url() -> String {
+    let config = get_config();
+    format!("{}{}", config.common.web_url, config.common.base_uri)
+}
+
+/// Shortens the given original URL and stores it in the database
+pub async fn shorten(original_url: &str) -> Result<String, anyhow::Error> {
+    let mut short_id = md5::short_hash(original_url);
+
+    // Check if the generated short_id is already present in the database
+    if db::short_url::get(short_id.as_str()).await.is_ok() {
+        // Handle hash conflict - create a new hash using the timestamp
+        let timestamp = Utc::now().timestamp();
+        let input = format!("{}{}", original_url, timestamp);
+        short_id = md5::short_hash(&input);
+    }
+
+    let entry = ShortUrlRecord::new(short_id.clone(), original_url.to_string());
+
+    // Store the short_id and original_url in the database
+    let result = db::short_url::set(&short_id, entry.clone()).await;
+
+    match result {
+        Ok(_) => Ok(format!("{}/short/{}", get_base_url(), entry.short_id)),
+        Err(e) => {
+            // Try to downcast the error to `infra::errors::Error`
+            if let Some(infra_error) = e.downcast_ref::<infra::errors::Error>() {
+                match infra_error {
+                    Error::DbError(DbError::UniqueViolation) => {
```

Comment:
Here, we have no `unique` index for `original_url`, so, if it return error `uniqueViolation` it shouldn't get original_url and return the short_url.